### PR TITLE
[WIP] Add config module for querying PacRunner

### PR DESCRIPTION
**Pull request by klausenbusk**
Credit: David Woodhouse David.Woodhouse@intel.com

Just to keep track of it, not tested or anything yet.

### CODE REVIEW

File: libproxy/modules/config_pacrunner.cpp [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+			throw runtime_error("Unrecognised proxy response from PacRunner: " + string(str));
+		}
+		return response;
+	}
+
+private:
+	DBusConnection *conn;
+};
+
+MM_MODULE_INIT_EZ(pacrunner_config_extension, true, NULL, NULL);
```
Comment (DimStar77) [inline libproxy/modules/config_pacrunner.cpp]: That sounds a bit generic to be an always valid config module... makes the module apparently pick up the config even if pacrunner is not being used.

The 'make thus module valid' test should at least query if pacrunner listens on the dbus... if not, it's very likely not in use.

@npmccallum thoughts?

**Comment by dwmw2** [comment]
Thanks for filing this here; I'd filed it in code.google.com a long time ago but had neglected to chase it up again. Checking for validity certainly seems like a sane request. Perhaps it can be as simple as a function which *tries* to request proxies from PacRunner for http://localhost/ and return true if it gets any answer; false if it fails?

**Comment by DimStar77** [comment]
I'd say the function should purely check if PacRunner runs on the dbus... if not it's not there and the module thus should not be used

**Comment by dwmw2** [comment]
Given that PacRunner can be instantiated upon request... is there a way to check if it exists on the bus, without actually making a request to it?

**Comment by cgwalters** [comment]
I think this type of case is why the bus implements `ListActivatiableNames` (try e.g. `busctl --activatable`).

**Comment by dwmw2** [comment]
Hm... PacRunner provides two interfaces:
 • org.pacrunner.Manager for NetworkManager/ConnMan/etc. to tell it about the configuration
 • org.pacrunner.Client for clients to use.

Here, of course, libproxy is only using the latter.

What we actually want is for automatic activation to happen only for calls to org.pacrunner.Manager, and *not* for calls to org.pacrunner.Client. That is, when somebody *tells* it the config, it comes to life. But not just because somebody *asks* it.

Is there a way to achieve that? Both interfaces are under the org.pacrunner service... do we need to split them? Discussion probably best taken to the connman mailing list (which is where pacrunner is maintained)...

**Comment by cgwalters** [comment]
I don't think in a backwards compatible way.  However, it is possible for clients to use [a flag to disable auto-activation](https://dbus.freedesktop.org/doc/dbus-specification.html#message-protocol).  So call `dbus_message_set_auto_start(msg, FALSE)` in the libproxy code?

**Comment by dwmw2** [comment]
OK, thanks. That probably makes sense — although it's only really useful if *all* the clients do it. If just one of them activates a PacRunner dæmon that nobody configures, that's suboptimal.

Back to libproxy though... we still end up with the problem that we want our *check* function (currently just `true` to make the connection and it might end up with a `DBusConnection *` that we want to keep... but we have nowhere to put it yet.

**Comment by klausenbusk** [comment]
I'm going to close this, I found another solution for my use-case. If someone else want to finish it, feel free to fork my branch :)

**Pull request closed without merging**